### ISSUE TITLE
New check: Blank line after type definitions

### ISSUE THREAD

**Issue by matrei** [checkstyle/checkstyle#17393]
**Example of desired behavior**

```bash
/var/tmp $ javac Example.java

/var/tmp $ cat config.xml
<module name="BlankLineAfterTypeCheck"/>

/var/tmp $ cat Example.java
public class Example {
    private int a;
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -c config.xml Example.java
[ERROR] /path/to/Example.java:1:1: Blank line required after type definition opening brace [BlankLineAfterTypeCheck]
```

I would like to be able to enforce a blank line after type definitions.
Empty type bodies would not need a blank line.

I will be willing to try and contribute this functionality if accepted.

**Related issues**

- #5439
- #11286

**Comment by romani** [comment]
I read previous tickets  on this, and more inclined to new Check that does only one validation and nothing else.

a side note:
There are bunch of discussions and ideas on when to place empty line after method declaration, it is not that simple. So it gives one more prove that there should be no Check that works for all, but better to have numerous Checks that cover special cases.

I am ok to make BlankLineAfterTypeCheck .

@matrei , please share examples to all types: records, interfaces, enums
please share examples on what is good code and and is not good code according to Check. You will reuse such examples during test creation for sure, it is not waste of time for description.

@nrmancuso , please review and approve issue.

**Comment by nrmancuso** [comment]
I am good to approve one we have established the expected behavior outlined in the additional examples that @romani has requested. I believe that the proposed name could also use some improvement, since it sounds like we are checking for a blank line after the entire type definition instead of checking for a blank line after the opening brace.

**Comment by romani** [comment]
@nrmancuso , what are your suggestions on name of Check ?
I did not come up with good user friendly name.

**Comment by nrmancuso** [comment]
How about `ClassBodyPaddingCheck`? I think we could reasonably extend such a check to include an option for the closing brace as well.

**Comment by romani** [comment]
Good idea , I like `ClassBodyPadding`  we had similar requests for empty lines at the beginning and end
https://github.com/checkstyle/checkstyle/issues/5440
https://github.com/checkstyle/checkstyle/issues/5439

**Comment by matrei** [comment]
Thanks for getting back so fast!
Here is a swing at a more complete specification.

## TypeBodyPadding
This check will enforce **_at least one_** blank (only whitespace) line immediately after the opening brace and/or before the closing brace of type definitions, such as `class`, `interface`, `enum`, and `record`.

### 🔧 Configuration
```xml
<module name="TypeBodyPadding">
    <property name="starting" value="true" /><!-- Require a blank line after the opening brace. default: true -->
    <property name="ending" value="true" /><!-- Require a blank line before the closing brace. default: true -->
    <property name="allowEmpty" value="true" /><!-- Don't enforce padding for completely empty types. default: true -->
</module>
```

### ✅ Allowed
```java
public class Allowed1 {

    private int a;

}

public class Allowed2
{

    private int a;

}

public class Allowed3 {
} // empty class — allowed

public class Allowed4 {} // empty class — allowed

public class Allowed5
{} // empty class — allowed

public class Allowed6
{
} // empty class — allowed

public class Allowed7 {

} // empty class — blank line allowed
```

### ❌ Disallowed
```java
public class Disallowed1 {
    private int a;
} // missing padding start and end

public class Disallowed2 {
    private int a;

} // missing padding start

public class Disallowed3 {

    private int a;
} // missing padding end

public class Disallowed4
{
    private int a;
} // missing padding start and end

public class Disallowed5
{
    private int a;

} // missing padding start

public class Disallowed6 {
    // This is field a
    private int a;

} // comments are not a blank line, missing padding start
```

... and the same for Interfaces, Records and Enums

**Comment by nrmancuso** [comment]
Sold

**Comment by MiniPiku** [comment]
Hi @romani,

I see that this issue was previously approved and the expected behavior was clarified, but there has not been recent activity.

If it is still open for contribution, I would like to work on implementing `TypeBodyPaddingCheck`.

Here is my initial draft design proposal:

Scope
- The check will validate blank line padding inside type definitions:
  - class
  - interface
  - enum
  - record

Configuration
- starting (default: true)
  Require at least one blank (whitespace-only) line immediately after the opening brace.

- ending (default: true)
  Require at least one blank (whitespace-only) line immediately before the closing brace.

- allowEmpty (default: true)
  Skip validation for completely empty type bodies.

Implementation Approach
- Register tokens:
  CLASS_DEF, INTERFACE_DEF, ENUM_DEF, RECORD_DEF

- For each type:
  1. Locate OBJBLOCK.
  2. Retrieve LCURLY and RCURLY.
  3. If allowEmpty = true and body contains no members, skip.
  4. For starting:
     - Find first meaningful child after LCURLY.
     - Validate that at least one full whitespace-only line exists between LCURLY and that element.
     - Use FileContents to ensure the line is truly blank (comments should not count).
  5. For ending:
     - Find last meaningful child before RCURLY.
     - Validate that at least one whitespace-only line exists before RCURLY.

Edge Cases to Consider
- One-line empty types: class A {}
- Types with comments immediately after '{'
- Nested types
- Multiple blank lines (should be allowed)
- Annotations inside type bodies

If this direction looks good, I will start with a draft PR including minimal functionality and expand with full test coverage afterward.

Please let me know if any adjustments are preferred before I begin implementation.

**Comment by Adam-Gamil** [comment]
Hi @romani, I am currently preparing for GSoC and would love to take this over and try to finish it. I will review the work done in the previous PR and open a fresh Draft PR with the implementation as soon as possible.

**Comment by romani** [comment]
There is abandoned PR, please reuse code from it

**Comment by Adam-Gamil** [comment]
> There is abandoned PR, please reuse code from it

Sure i will

**Comment by Atharv3221** [comment]
@Adam-Gamil are you working on this ??

**Comment by Adam-Gamil** [comment]
> [@Adam-Gamil](https://github.com/Adam-Gamil) are you working on this ??

@Atharv3221 Yes i am working on it

**Comment by Adam-Gamil** [comment]
Hi @romani,

I am currently reviewing the code from the abandoned PR while i am working on the check and want to make sure I understand before I push the new Draft PR. I want to clarify a few things:

1. **`skipInner` and `skipLocal` Properties**: The previous contributor added these properties to let users toggle whether the check applies to nested/local types. Should I keep these as configurable properties in the API, or should the Check just enforce the padding rules universally on all type definitions (top-level, inner, and local) to keep it simple?

2. **Test Diversity**: I am building test inputs that cover `record`, `enum`, and ` interface` types. Should these be added as separate input files each testing one type, or combined into single files showing multiple type kinds together, or or a mix of both?.

3. **General direction**: Are there any other specific requirements or corrections you'd like addressed while I am working on the check ?

Please let me know your preference for Point 1.
Thanks in advance

**Comment by Atharv3221** [comment]
> Hi @romani,
> 
> I am currently reviewing the code from the abandoned PR while i am working on the check and want to make sure I understand before I push the new Draft PR. I want to clarify a few things:
> 
> 1. **`skipInner` and `skipLocal` Properties**: The previous contributor added these properties to let users toggle whether the check applies to nested/local types. Should I keep these as configurable properties in the API, or should the Check just enforce the padding rules universally on all type definitions (top-level, inner, and local) to keep it simple?
> 
> 2. **Test Diversity**: I am building test inputs that cover `record`, `enum`, and ` interface` types. Should these be added as separate input files each testing one type, or combined into single files showing multiple type kinds together, or or a mix of both?.
> 
> 3. **General direction**: Are there any other specific requirements or corrections you'd like addressed while I am working on the check ?
> 
> Please let me know your preference for Point 1.
> Thanks in advance

Yes for one
For second yes you can use multiple types in a single file just don't make the file too complex

**Comment by matrei** [comment]
> There is abandoned PR, please reuse code from it 

It's not abandoned, I just ran out of time to finish it at that particular time, and I am planning to pick it up when I get a chance. That being said, I welcome any efforts to finish it.

**Comment by Adam-Gamil** [comment]
Sorry for responding late

First, thank you @Atharv3221 for the advice. Really helped me.

Second, thank you too @matrei  for starting the original PR for this check. I know how hard it is to start a new rule from scratch, I basically took your old code and continued working on it. I updated the logic to pass the tests coverage rules, added a few more test cases and refactored some of the original code of yours.

Since this was your idea first, I would be really happy if you took a look at the code to see if it does what you wanted.

The check passed all the local tests of `mvn clean verify`, waiting for CI tests

Looking forward to your feedback!

**Comment by mahfouz72** [comment]
@matrei @Adam-Gamil @Atharv3221 Since all of you proposed starting or already started this check, does anyone have the energy and time to complete it now? It is a highly requested check.

**Comment by Atharv3221** [comment]
@Adam-Gamil are you working on this

**Comment by Adam-Gamil** [comment]
@Atharv3221 sorry for late reply, no i stoped working on it
you can take it

**Comment by romani** [comment]
fix is merged, temp link to doc is at https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/2cc5cf7_20260717172520/checks/whitespace/typebodypadding.html#
